src/main/java/org/folio/auth/authtokenmodule/storage/RefreshTokenStore.java

```diff
@@ -0,0 +1,202 @@
+    //    consider the user's account compromised and revoke all their refresh tokens.
+    //    They will need to login again to get a new RefreshToken that isn't revoked
+    //    or redeemed.
+    UUID tokenId = refreshToken.getId();
+    UUID userId = refreshToken.getUserId();
+    String table = tableName(tenant, REFRESH_TOKEN_SUFFIX);
+
+    log.info("Checking token redeemed of token id {} for tenant {}", tokenId, tenant);
+
+    String select = "SELECT is_redeemed, is_revoked FROM " + table + "WHERE id=$1";
+    Tuple where = Tuple.of(tokenId);
+
+    return pool.withConnection(conn -> {
+      return getRow(conn, select, where).compose(row -> {
+        Boolean isRedeemed = row.getBoolean("is_redeemed");
+        Boolean isRevoked = row.getBoolean("is_revoked");
+
+        log.info("Redeemed status of {} token id {} is {}", REFRESH_TOKEN_SUFFIX, tokenId, isRedeemed);
+        log.info("Revoked status of {} token id {} is {}", REFRESH_TOKEN_SUFFIX, tokenId, isRevoked);
+
+        // The token could have been revoked by a different request so we check that first.
+        // If it has been revoked, all tokens should already have been revoked so we can safely
+        // return.
+        if (isRevoked) {
+          log.info("Token {} has been revoked", tokenId);
+          return Future.failedFuture("Token is revoked");
+        }
+
+        if (!isRedeemed) {
+          // Token has not been used more than once. Set it as redeemed so it can't be used again
+          // and return a success.
+          log.info("Token {} has not yet been redeemed so it is not revoked", tokenId);
+          return setTokenRedeemed(conn, refreshToken);
+        }
+
+        // isRedeemed is true so the token has been used more than once. Revoke
+        // all tokens for the user.
+        String leakedMessage = "Refresh token {} attempted to be used twice." +
+        " It is considered leaked. Revoking all tokens for user {}.";
+        log.info(leakedMessage, tokenId, userId);
+
+        return revokeAllTokensForUser(conn, userId).compose(y -> {
+          return Future.failedFuture("Token leaked. Revoked all tokens for user.");
+        });
+      });
+    });
+  }
+
+  /**
+   * Cleans up (deletes) refresh tokens which have passed their time of expiration.
+   * This method should be called periodically by a timer mechanism which can be any
+   * reasonable amount of time to keep things fresh but not cause undue system load.
+   * Clients need to ensure that they are requesting new refresh tokens before they
+   * expire.
+   * @return A failed future should the deletion fail. A succeeded future if the
+   * deletion succeeds.
+   */
+  public Future<Void> cleanupExpiredTokens() {
+    long now = Instant.now().getEpochSecond();
+    log.info("Cleaning up tokens which are older than: {}", now);
+
+    String delete = "DELETE FROM " + tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+      "WHERE expires_at<$1";
```

Comment:
`DELETE FROM ... WHERE id IN (SELECT id FROM ... WHERE expires_at<$1 FOR UPDATE SKIP LOCKED)`

This skips rows that are locked and therefore doesn't wait until they get unlocked.
This locks rows to be deleted and therefore avoids rollbacks.